Assets/Scripts/StandardSamples/Discord/DiscordManager.cs

```diff
@@ -121,13 +123,54 @@ private class TokenReader
         {
             public string access_token;
             public float expires_in;
+            public string refresh_token;
         }
 
         private string cachedToken = null;
+        private string refresh_token = null;
         private DateTime tokenExpiration;
 
         private event Action<string> onTokenReceived;
 
+        public void StartConnectLoginWithDiscord(EOSManager.OnConnectLoginCallback onLoginCallback)
+        {
+            StartConnectLoginWithDiscord(onLoginCallback, false);
+        }
+
+        private void StartConnectLoginWithDiscord(EOSManager.OnConnectLoginCallback onLoginCallback, bool isRetry = false)
+        {
+            // First acquire an auth token
+            // If it's invalid, end here. If a valid was returned, attempt to login.
+            // If the result code is exactly Result.ConnectExternalTokenValidationFailed, then reacquire the token and try again
+            // Then otherwise if the login succeeds or fails, call the onLoginCallback with the result.
+
+            Discord.DiscordManager.Instance.RequestOAuth2Token((string returnedToken) =>
+            {
+                // Without a token, end here
+                if (string.IsNullOrEmpty(returnedToken))
+                {
+                    onLoginCallback?.Invoke(new Epic.OnlineServices.Connect.LoginCallbackInfo() { ResultCode = Epic.OnlineServices.Result.InvalidAuth });
+                    return;
+                }
+
+                // Attempt to login to Discord
+                EOSManager.Instance.StartConnectLoginWithOptions(ExternalCredentialType.DiscordAccessToken, returnedToken, onloginCallback: (LoginCallbackInfo callbackInfo) =>
+                {
+                    if (!isRetry && callbackInfo.ResultCode == Result.ConnectExternalTokenValidationFailed)
+                    {
+                        // If the auth failed for exactly this reason, then try logging in again, doing the whole process again to acquire a new token
+                        cachedToken = null;
+                        StartConnectLoginWithDiscord(onLoginCallback, true);
+                    }
```

Comment:
As a general principle, when I have a bit of code that needs to "retry" a connection, I like to structure it in such a way that the number of retries is variable, so that in the future if more than one retry is called for, that's easy to scale.

In my experience, this makes for cleaner, clearer code. Additionally, implementation of that "retrying" ought to be accomplished via a helper method that makes it explicit that a retry will be attempted.

This is more a thought than an explicit critique; I think it's an idea worth considering, but not a concept upon which I would ask for changes to be made.

Reply:
I find this agreeable. These boolean retry arguments are being traded for ints that count the number of retries, and a const int inside the function that gates how many times to retry.